controllers/etcdrestore_controller.go

```diff
@@ -0,0 +1,524 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-logr/logr"
+	etcdv1alpha1 "github.com/improbable-eng/etcd-cluster-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// EtcdRestoreReconciler reconciles a EtcdRestore object
+type EtcdRestoreReconciler struct {
+	client.Client
+	Log             logr.Logger
+	RestorePodImage string
+}
+
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdrestores,verbs=get;list;watch
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdrestores/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=persistantvolumeclaims/status,verbs=get;watch;update;patch;create
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;watch;list;create
+
+func name(o metav1.Object) types.NamespacedName {
+	return types.NamespacedName{
+		Namespace: o.GetNamespace(),
+		Name:      o.GetName(),
+	}
+}
+
+const (
+	restoreContainerName = "etcd-restore"
+	restoredFromLabel    = "etcd.improbable.io/restored-from"
+	restorePodLabel      = "etcd.improbable.io/restore-pod"
+)
+
+func markPVC(restore etcdv1alpha1.EtcdRestore, pvc *corev1.PersistentVolumeClaim) {
+	if pvc.Labels == nil {
+		pvc.Labels = make(map[string]string)
+	}
+	pvc.Labels[restoredFromLabel] = restore.Name
+}
+
+func IsOurPVC(restore etcdv1alpha1.EtcdRestore, pvc corev1.PersistentVolumeClaim) bool {
+	return pvc.Labels[restoredFromLabel] == restore.Name
+}
+
+func markPod(restore etcdv1alpha1.EtcdRestore, pod *corev1.Pod) {
+	if pod.Labels == nil {
+		pod.Labels = make(map[string]string)
+	}
+	pod.Labels[restoredFromLabel] = restore.Name
+	pod.Labels[restorePodLabel] = "true"
+}
+
+func IsOurPod(restore etcdv1alpha1.EtcdRestore, pod corev1.Pod) bool {
+	return pod.Labels[restoredFromLabel] == restore.Name &&
+		pod.Labels[restorePodLabel] == "true"
+}
+
+func markCluster(restore etcdv1alpha1.EtcdRestore, cluster *etcdv1alpha1.EtcdCluster) {
+	if cluster.Labels == nil {
+		cluster.Labels = make(map[string]string)
+	}
+	cluster.Labels[restoredFromLabel] = restore.Name
+}
+
+func (r *EtcdRestoreReconciler) podForRestore(restore etcdv1alpha1.EtcdRestore, pvc *corev1.PersistentVolumeClaim) *corev1.Pod {
```

Comment:
Is this used? I see there's also a free function `podForRestore` which is used.

Reply:
Oh this got left over. Deleting.